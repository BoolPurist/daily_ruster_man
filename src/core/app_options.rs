/// Used in the core modules to alter certain behaviors of app
use crate::cli::app_args::{DebugArgs, GenerellArgs, CliArgs};
#[derive(Getters)]
pub struct AppOptions {
    /// During development, this will have some value.
    debug: Option<DebugArgs>,
    #[getset(get = "pub")]
    general: GenerellArgs,
}

impl AppOptions {
    #[cfg(debug_assertions)]
    pub fn new(args: &CliArgs) -> Self {
        Self {
            general: args.args().clone(),
            debug: Some(args.debug_args().clone()),
        }
    }

    #[cfg(not(debug_assertions))]
    pub fn new(args: &CliArgs) -> Self {
        Self {
            general: args.args().clone(),
            debug: None,
        }
    }

    /// If true is return then the user local files are to be used even during development
    pub fn use_prod_local_share(&self) -> bool {
        if let Some(use_it) = &self.debug {
            use_it.user_local_share()
        } else {
            false
        }
    }
}
