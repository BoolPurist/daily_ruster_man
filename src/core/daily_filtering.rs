use super::DailyName;
use std::path::Path;

pub fn filter_out_non_daily<'a>(
    to_filter: impl Iterator<Item = &'a str> + 'a,
) -> impl Iterator<Item = DailyName> + 'a {
    to_filter.filter_map(|file_name| file_name.parse().ok())
}

pub fn strip_expect_file_name<T>(paths: &[T]) -> impl Iterator<Item = &str>
where
    T: AsRef<Path>,
{
    paths.iter().map(|full_path| {
        full_path
            .as_ref()
            .file_name()
            .expect("Could not get file name")
            .to_str()
            .expect("Could convert os string to utf string")
    })
}

#[cfg(test)]
mod testing {
    use chrono::NaiveDate;

    use super::*;
    use std::path::Path;
    #[test]
    fn test_strip_expect_file_name() {
        let given = vec![
            Path::new("/home/bla/text.txt"),
            Path::new("/home/bla/aaa.a"),
            Path::new("/home/bla/ssss.s"),
        ];

        let expected = vec!["text.txt", "aaa.a", "ssss.s"];

        let actual: Vec<&str> = strip_expect_file_name(&given).collect();

        assert_eq!(expected, actual);
    }

    #[test]
    fn test_filter_out_non_daily() {
        let given = ["text.txt", "2022_02_2_daily.md", "2001_2_22_daily.md"];
        let actual: Vec<NaiveDate> = filter_out_non_daily(given.into_iter())
            .map(|daily_name| daily_name.date())
            .collect();

        assert_eq!(
            actual,
            vec![
                NaiveDate::from_ymd_opt(2022, 2, 2).unwrap(),
                NaiveDate::from_ymd_opt(2001, 2, 22).unwrap()
            ]
        );
    }
}
