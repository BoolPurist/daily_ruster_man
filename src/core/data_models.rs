#[derive(Debug)]
pub enum PastFuture {
    Past(u32),
    Future(u32),
}
impl PastFuture {
    pub fn new(past_or_future: i32) -> Self {
        if past_or_future < 0 {
            Self::Past(past_or_future.unsigned_abs())
        } else {
            Self::Future(past_or_future.unsigned_abs())
        }
    }
}
#[derive(new, CopyGetters, Debug)]
#[getset(get_copy = "pub")]
pub struct DayOfYear {
    year: u32,
    day_of_year: u32,
}
#[derive(new, CopyGetters, Debug)]
#[getset(get_copy = "pub")]
pub struct DayMonthYear {
    year: u32,
    month: u32,
    day: u32,
}

#[derive(new, CopyGetters, Debug)]
#[getset(get_copy = "pub")]
pub struct FilterParamsYmD {
    year: Option<u32>,
    month: Option<u32>,
    day: Option<u32>,
}
