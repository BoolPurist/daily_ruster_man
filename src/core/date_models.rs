use crate::prelude::*;
#[derive(Debug)]
pub enum OpenByDaysInTime {
    Past(u32),
    Future(u32),
}

impl OpenByDaysInTime {
    pub fn new(past_or_future: i32) -> Self {
        if past_or_future < 0 {
            Self::Past(past_or_future.unsigned_abs())
        } else {
            Self::Future(past_or_future.unsigned_abs())
        }
    }
}

#[derive(new, CopyGetters, Debug)]
#[getset(get_copy = "pub")]
pub struct OpenByDayOfYear {
    year: u32,
    day_of_year: u32,
}
#[derive(new, CopyGetters, Debug)]
#[getset(get_copy = "pub")]
pub struct OpenByDayMonthYear {
    year: u32,
    month: u32,
    day: u32,
}

#[derive(Debug)]
pub enum OpenByMonthInYear {
    CurrentMonth,
    InCurrentYear(u32),
    WithYear { month: u32, year: u32 },
}

#[derive(CopyGetters, Debug)]
#[getset(get_copy = "pub")]
pub struct FindByYearMonthDay {
    year: Option<u32>,
    month: Option<u32>,
    day: Option<u32>,
}

pub enum FindByMonthInYear {
    All,
    InCurrentYear(u32),
    MonthYear { month: u32, year: u32 },
}

impl FindByYearMonthDay {
    pub fn new(y_opt: Option<u32>, m_opt: Option<u32>, d_opt: Option<u32>) -> AppResult<Self> {
        match (y_opt, m_opt, d_opt) {
            (Some(y), Some(m), Some(d)) => {
                if chrono::NaiveDate::from_ymd_opt(y as i32, m, d).is_none() {
                    bail!(
                        "year, month or day is not valid for given date: year: {y}, month: {m}, day: {d}"
                    )
                }
            }
            (_, Some(m), Some(d)) => {
                check_for_day(d)?;
                check_for_month(m)?;
            }
            (_, Some(m), None) => {
                check_for_month(m)?;
            }
            (_, None, Some(d)) => {
                check_for_day(d)?;
            }
            _ => (),
        };

        Ok(Self {
            year: y_opt,
            month: m_opt,
            day: d_opt,
        })
    }
}

pub fn check_for_month(m: u32) -> AppResult {
    const LOWER_BOUND: u32 = 1;
    const UPPER_BOUND: u32 = 12;

    if !(LOWER_BOUND..=UPPER_BOUND).contains(&m) {
        bail!("Month must be between {LOWER_BOUND} and {UPPER_BOUND}.")
    }

    Ok(())
}

pub fn check_for_day(d: u32) -> AppResult {
    const LOWER_BOUND: u32 = 1;
    const UPPER_BOUND: u32 = 31;

    if !(LOWER_BOUND..=UPPER_BOUND).contains(&d) {
        bail!("Day of month must be between {LOWER_BOUND} and {UPPER_BOUND}.")
    }

    Ok(())
}
