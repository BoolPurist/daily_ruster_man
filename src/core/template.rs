pub mod command_processor;

pub use command_processor::{CommandToExecute, OsCommandProcossor};

use std::{borrow::Cow, collections::HashMap};

use regex::{Regex, Replacer};

use self::command_processor::CommandProcessor;
use crate::prelude::*;

#[derive(Debug)]
pub enum PlaceholderTemplate<'a, T> {
    DirectValue(&'a str),
    Commmand(CommandToExecute<'a, T>),
}
pub fn replace_template_placeholders<'m, 't, T>(
    template: &'t str,
    placeholders: &'m mut HashMap<&'t str, PlaceholderTemplate<'t, T>>,
) -> Cow<'m, str>
where
    T: CommandProcessor,
{
    lazy_static! {
        static ref REGEX_PLACE_HOLDERS: Regex = Regex::new(r"(?mi)\{\s*(\S+?)\s*\}")
            .unwrap_or_else(|error| panic!(
                "Invalid expression inside function: {}\n. Reason: {:?}",
                stringify!(replace_template_placeholders),
                error
            ));
    }

    let mut found_errors_for_commmand: HashMap<String, String> = HashMap::new();
    let replacer = PlaceHolderReplacer::new(placeholders, &mut found_errors_for_commmand);

    let replacement = REGEX_PLACE_HOLDERS.replace_all(template, replacer);

    for error in found_errors_for_commmand.into_iter() {
        error!(
            "For key {0} the command could not be executed without errors.\n Error: {1}",
            error.0, error.1,
        )
    }
    replacement
}

pub struct PlaceHolderReplacer<'m, 'kv, T>
where
    T: CommandProcessor,
{
    map: &'m mut HashMap<&'kv str, PlaceholderTemplate<'kv, T>>,
    errors: &'m mut HashMap<String, String>,
}

impl<'m, 'kv, T> PlaceHolderReplacer<'m, 'kv, T>
where
    T: CommandProcessor,
{
    pub fn new(
        map: &'m mut HashMap<&'kv str, PlaceholderTemplate<'kv, T>>,
        errors: &'m mut HashMap<String, String>,
    ) -> Self {
        Self { map, errors }
    }
}

impl<'m, 'kv, T> Replacer for PlaceHolderReplacer<'m, 'kv, T>
where
    T: CommandProcessor,
{
    fn replace_append(&mut self, caps: &regex::Captures<'_>, dst: &mut String) {
        let full_match = caps
            .get(0)
            .expect("Unexpected: no full caputure group found even with a found match");
        let inner_match = caps.get(1).expect(
            "Unexpexted: no innner capture group found for matching placholders in a template",
        );

        let key = inner_match.as_str();

        match self.map.get_mut(key) {
            Some(to_insert) => match to_insert {
                PlaceholderTemplate::DirectValue(direct_insert) => {
                    debug!(
                        "Inserted for placeholder key {} direct value {}",
                        key, direct_insert
                    );
                    dst.push_str(direct_insert);
                }
                PlaceholderTemplate::Commmand(command_to_insert) => {
                    let command_output = command_to_insert.get_std_out();

                    dst.push_str(command_output);

                    debug!(
                        "Inserted for placeholder key {} the output {}",
                        key, command_output,
                    );

                    let command_error = command_to_insert.get_std_err();
                    if let Some(stderr) = command_error {
                        if !self.errors.contains_key(key) {
                            debug!("Found error output for key {}. Output: {}", key, stderr);
                            self.errors.insert(key.to_owned(), stderr.to_owned());
                        }
                    }
                }
            },
            None => dst.push_str(full_match.as_str()),
        }
    }
}

fn parse_commmand_text(to_parse: &str) -> AppResult<Vec<String>> {
    shellwords::split(to_parse).map_err(AppError::new)
}

#[cfg(test)]
pub fn return_dummy_processed_command(input: &str) -> String {
    let splitted = parse_commmand_text(input).expect("Invalid command text given");
    let mut output = String::from("Executed_");
    output.push_str(&splitted.join("_"));
    output
}

#[cfg(test)]
mod testing {
    use std::collections::HashMap;
    use super::{replace_template_placeholders, CommandToExecute};
    use super::return_dummy_processed_command;
    use crate::core::template::PlaceholderTemplate;

    use super::command_processor::{MockCommandProcessor, CommandOutput};
    type FakeCommandOutput<'l> = PlaceholderTemplate<'l, MockCommandProcessor>;
    fn create_dummmy_command_processor<'a>(
        command_text: &'a str,
        expected_error: Option<String>,
        times: usize,
    ) -> FakeCommandOutput<'a> {
        let mut mock = MockCommandProcessor::new();
        let expected_error = expected_error.clone();
        mock.expect_process()
            .times(times)
            .returning(move |to_process| {
                CommandOutput::new(
                    return_dummy_processed_command(to_process),
                    expected_error.clone(),
                )
            });

        PlaceholderTemplate::Commmand(CommandToExecute::new_with(command_text, mock))
    }
    #[test]
    fn should_complement_template_with_command_processor() {
        let given_templa = r#"
# for {hello}
## Uptime is {how_long}
- Uptime is now {how_long_now}
## Uptime is {how_long} again
more {hello}
### Not {found}

Should inserted {how_long_error} even with errors
"#;
        let mut map: HashMap<&str, FakeCommandOutput> = HashMap::from_iter(
            vec![
                ("hello", PlaceholderTemplate::DirectValue("world")),
                (
                    "how_long",
                    create_dummmy_command_processor("uptime", None, 1),
                ),
                (
                    "not_to_be_inserted",
                    create_dummmy_command_processor("oh_oh", None, 0),
                ),
                (
                    "how_long_error",
                    create_dummmy_command_processor("uptime xxx", Some("error".to_string()), 1),
                ),
                (
                    "how_long_now",
                    create_dummmy_command_processor("uptime now", None, 1),
                ),
            ]
            .into_iter(),
        );

        // Act
        let actual = replace_template_placeholders(given_templa, &mut map);
        insta::assert_display_snapshot!(actual);
    }
}
