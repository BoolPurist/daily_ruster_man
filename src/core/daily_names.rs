use super::data_models::*;
use chrono::prelude::*;
use std::fmt::Display;
use std::str::FromStr;
use thiserror::Error;

const DAILY_INFIX: &str = "daily";
const DIGIT_SEP: &str = "_";
pub const MD_EXT: &str = "md";

#[derive(Debug, PartialEq, Eq)]
pub struct DailyName {
    date: NaiveDate,
    name: String,
}

impl DailyName {
    pub fn new(date: NaiveDate, ext: &str) -> Self {
        let year = date.year();
        let month = date.month();
        let day = date.day();
        let name = Self::to_format(year, month, day, ext);

        Self { name, date }
    }

    fn to_format(year: i32, month: u32, day: u32, ext: &str) -> String {
        format!("{year}{DIGIT_SEP}{month:02}{DIGIT_SEP}{day:02}{DIGIT_SEP}{DAILY_INFIX}.{ext}",)
    }

    pub fn get_name(&self) -> &str {
        &self.name
    }
    pub fn get_date(&self) -> NaiveDate {
        self.date
    }
    pub fn create_daily_name_from(date: NaiveDate) -> Self {
        Self::new(date, MD_EXT)
    }
    pub fn create_today_name() -> DailyName {
        let now = chrono::Local::now();
        let date_now = now.date_naive();

        Self::create_daily_name_from(date_now)
    }

    fn create_from_point_and_range(range: &PastFuture, mut wanted_date: NaiveDate) -> Self {
        match range {
            PastFuture::Past(days) => wanted_date -= chrono::Duration::days(*days as i64),
            PastFuture::Future(days) => wanted_date += chrono::Duration::days(*days as i64),
        };

        Self::create_daily_name_from(wanted_date)
    }

    pub fn create_from_range(range: &PastFuture) -> Self {
        let wanted_date = chrono::Local::now().date_naive();
        Self::create_from_point_and_range(range, wanted_date)
    }
}

impl Display for DailyName {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.name)
    }
}

impl FromStr for DailyName {
    type Err = ParseDailyNameError;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut splits = s.split(DIGIT_SEP);

        let (y_parsed, m_parsed, d_parsed) = match (splits.next(), splits.next(), splits.next()) {
            (Some(year), Some(month), Some(day)) => {
                let parsed_year: i32 = year.parse().or(Err(ParseDailyNameError::YearInvalid))?;
                let parsed_month: u32 = month.parse().or(Err(ParseDailyNameError::MonthInvalid))?;
                let parsed_daiy: u32 = day.parse().or(Err(ParseDailyNameError::DayInvalid))?;

                Ok((parsed_year, parsed_month, parsed_daiy))
            }
            (Some(_), Some(_), None) => Err(ParseDailyNameError::MissingMonth),
            (Some(_), None, None) => Err(ParseDailyNameError::MissingYear),
            _ => unreachable!(),
        }?;

        if y_parsed < 0 {
            return Err(ParseDailyNameError::YearInvalid);
        }

        let date = NaiveDate::from_ymd_opt(y_parsed, m_parsed, d_parsed)
            .ok_or(ParseDailyNameError::InvalidDate)?;

        let new_self = Self {
            name: s.to_owned(),
            date,
        };
        Ok(new_self)
    }
}

#[derive(Error, Debug, PartialEq, Eq)]
pub enum ParseDailyNameError {
    #[error("Invalid format for year")]
    YearInvalid,
    #[error("Invalid format for month")]
    MonthInvalid,
    #[error("Invalid format for day")]
    DayInvalid,
    #[error("No year provided for the name of daily")]
    MissingYear,
    #[error("No month provided for the name of daily")]
    MissingMonth,
    #[error("Date is not valid")]
    InvalidDate,
}

#[cfg(test)]
mod testing {
    use super::*;

    #[test]
    fn test_parse_error() {
        assert_if_parse_fails_with("daily.md", ParseDailyNameError::MissingYear);
        assert_if_parse_fails_with("2000_daily.md", ParseDailyNameError::MissingMonth);
        assert_if_parse_fails_with("aa_02_07_daily.md", ParseDailyNameError::YearInvalid);
        assert_if_parse_fails_with("2000_z_2_daily.md", ParseDailyNameError::MonthInvalid);
        assert_if_parse_fails_with("2000_2_-7_daily.md", ParseDailyNameError::DayInvalid);
        assert_if_parse_fails_with("2000_2_32_daily.md", ParseDailyNameError::InvalidDate);
        assert_if_parse_fails_with("2000_0_1_daily.md", ParseDailyNameError::InvalidDate);
        assert_if_parse_fails_with("-78_2_2_daily.md", ParseDailyNameError::YearInvalid);
    }

    #[test]
    fn test_parse() {
        assert_parse(
            "2023_01_07_daily.md",
            NaiveDate::from_ymd_opt(2023, 1, 7).unwrap(),
        );
        assert_parse(
            "1981_11_24_daily.md",
            NaiveDate::from_ymd_opt(1981, 11, 24).unwrap(),
        );
        assert_parse(
            "2001_02_20_daily.md",
            NaiveDate::from_ymd_opt(2001, 2, 20).unwrap(),
        );
    }

    #[test]
    fn test_create_daily_in_past() {
        assert_daily_from_range(PastFuture::Past(2), 2022, 2, 8, 2022, 2, 6);
        assert_daily_from_range(PastFuture::Past(40), 2022, 2, 7, 2021, 12, 29);
        assert_daily_from_range(PastFuture::Past(0), 1980, 1, 1, 1980, 1, 1);
    }
    #[test]
    fn test_create_daily_in_future() {
        assert_daily_from_range(PastFuture::Future(2), 2022, 2, 8, 2022, 2, 10);
        assert_daily_from_range(PastFuture::Future(40), 2022, 2, 7, 2022, 3, 19);
        assert_daily_from_range(PastFuture::Future(0), 1980, 1, 1, 1980, 1, 1);
    }

    #[test]
    fn test_new() {
        assert_daily_name_new(2000, 8, 20);
        assert_daily_name_new(1988, 5, 2);
        assert_daily_name_new(2022, 1, 30);
        assert_daily_name_new(1970, 12, 1);
    }

    fn assert_daily_from_range(
        range: PastFuture,
        g_year: u32,
        g_month: u32,
        g_day: u32,
        ex_year: u32,
        ex_month: u32,
        ex_day: u32,
    ) {
        let given_date =
            NaiveDate::from_ymd_opt(g_year as i32, g_month, g_day).expect("Invalid given date");
        let actual = DailyName::create_from_point_and_range(&range, given_date);
        let expected = DailyName::new(
            NaiveDate::from_ymd_opt(ex_year as i32, ex_month, ex_day)
                .expect("Invalid expected date"),
            MD_EXT,
        );
        assert_eq!(expected, actual);
    }

    fn assert_daily_name_new(y: i32, m: u32, d: u32) {
        let given = NaiveDate::from_ymd_opt(y, m, d).expect("actual date as test input is invalid");

        let expected = format!("{y}_{m:02}_{d:02}_daily.{MD_EXT}");

        let actual = DailyName::new(given, MD_EXT);
        assert_eq!(expected, actual.get_name());
    }

    fn assert_if_parse_fails_with(invalid_input: &str, expected_error: ParseDailyNameError) {
        match invalid_input.parse::<DailyName>() {
            Ok(_) => panic!("Should not parse on this invalid input {}", invalid_input),
            Err(error) => assert_eq!(error, expected_error),
        };
    }

    fn assert_parse(valid: &str, expected: NaiveDate) {
        if let Ok(parsed) = valid.parse::<DailyName>() {
            assert_eq!(parsed.date, expected);
        } else {
            panic!("Did parse valid input correctly, ({})", valid);
        }
    }
}
