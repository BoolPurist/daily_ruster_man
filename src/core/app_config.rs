pub mod path_from_config;
pub use path_from_config::PatchFromConfig;

use std::{
    path::{PathBuf, Path},
    collections::HashMap,
    borrow::Cow,
};

use serde::Deserialize;

use crate::core::template::{CommandToExecute, OsCommandProcossor};
use crate::prelude::*;

use super::{
    constants::CONF_FILE_NAME, template::PlaceholderTemplate, file_access, app_options::AppOptions,
    dates_names::ResolvePlaceholders,
};

macro_rules! path_from_conf_getter {
    ($field:ident) => {
        pub fn $field(&self) -> PatchFromConfig {
            PatchFromConfig::new(self.$field.clone())
        }
    };
}

#[derive(Deserialize, Default, Debug, Getters)]
/// Contains access to data which is provided by the conf file of app located at the app conf
/// folder
pub struct AppConfig {
    yearly_template: Option<String>,
    monthly_template: Option<String>,
    daily_template: Option<String>,
    #[getset(get = "pub")]
    data_foler: Option<String>,
    placeholders: Option<Vec<PlaceHolder>>,
    #[getset(get = "pub")]
    editor: Option<String>,
    #[serde(skip)]
    #[getset(get = "pub")]
    /// Path to folder where the config file loaded from
    root_path: PathBuf,
}

impl AppConfig {
    path_from_conf_getter! {monthly_template}
    path_from_conf_getter! {yearly_template}
    path_from_conf_getter! {daily_template}

    /// Fetches all placeholders from the config.
    pub fn create_template_placeholder_for<'a>(
        &'a self,
        journal: &impl ResolvePlaceholders,
    ) -> HashMap<&'_ str, PlaceholderTemplate<'_, OsCommandProcossor>> {
        return match &self.placeholders {
            None => HashMap::new(),
            Some(read_placeholders_from_config) => {
                let mut output: HashMap<&str, PlaceholderTemplate<'a, OsCommandProcossor>> =
                    HashMap::with_capacity(read_placeholders_from_config.len());
                for to_convert in read_placeholders_from_config {
                    let value = match to_convert.is_command() {
                        Some(is) => {
                            if *is {
                                PlaceholderTemplate::Commmand(CommandToExecute::new(
                                    to_convert.value(),
                                ))
                            } else {
                                resolve_direct_value(to_convert, journal)
                            }
                        }
                        None => resolve_direct_value(to_convert, journal),
                    };
                    output.insert(to_convert.key().as_str(), value);
                }

                output
            }
        };

        fn resolve_direct_value<'a>(
            to_convert: &'a PlaceHolder,
            journal: &impl ResolvePlaceholders,
        ) -> PlaceholderTemplate<'a, OsCommandProcossor> {
            use crate::core::constants::{PREFIX_FOR_BUITLIN_VAR, SUFFIX_FOR_BUITLIN_VAR};

            let to_convert = to_convert.value().as_str();
            let trimmed = to_convert
                .trim_start_matches(PREFIX_FOR_BUITLIN_VAR)
                .trim_end_matches(SUFFIX_FOR_BUITLIN_VAR);

            let resolved = journal.resolve_variable(trimmed);

            // PREFIX_FOR_BUITLIN_VAR, left, and SUFFIX_FOR_BUITLIN_VAR, right was removed.
            // Use the intial ref with left and right not trimmed away
            if let Cow::Borrowed(_trimmed) = resolved {
                PlaceholderTemplate::DirectValue(to_convert.into())
            } else {
                PlaceholderTemplate::DirectValue(resolved)
            }
        }
    }

    pub fn try_from_file_system(option: &AppOptions) -> AppResult<Option<Self>> {
        let resolved_path = if let Some(path) = option.general().config_path() {
            debug!("Using  conf path provided by cli or env.");
            file_access::resolve_str_as_path(path)
        } else {
            debug!("Using conf path provided by os.");
            let path = file_access::fetch_path_conf(option)?;
            file_access::resolve_path(&path)
        };
        debug!("Using {:?} as conf folder for app.", &resolved_path);

        let path_to_conf_file = resolved_path.join(CONF_FILE_NAME);

        debug!("Using {:?} as conf file for app.", &path_to_conf_file);
        if path_to_conf_file.exists() {
            let content = std::fs::read_to_string(&path_to_conf_file)
                .context("could not read config file by given path")?;

            match toml::from_str::<AppConfig>(&content) {
                Ok(mut parsed_content) => {
                    parsed_content.root_path = resolved_path;
                    Ok(Some(parsed_content))
                }
                Err(error) => {
                    warn!("App config file is not in valid format.\n Error: {}", error);
                    Ok(None)
                }
            }
        } else {
            info!("No config file found at {:?}", path_to_conf_file);
            Ok(None)
        }
    }

    /// Returns none if there is no template file at the given parameter.
    pub fn try_get_template_file_content(&self, path: &Path) -> AppResult<Option<String>> {
        if path.exists() {
            let template_content = std::fs::read_to_string(path)?;
            info!("Template path found at {:?}", path);
            Ok(Some(template_content))
        } else {
            info!("No template found at {:?}", path);
            Ok(None)
        }
    }
}

#[derive(Deserialize, Getters, Debug)]
#[getset(get = "pub")]
/// Represents a given placeholder from the user to augment a template with costum values or output
/// of a given command.
/// Placeholder are currrenlty retrieved from the config file.
pub struct PlaceHolder {
    /// which is searched for in the template
    key: String,
    /// by which the key in the template is replaced
    value: String,
    /// If true instead treat `value` as command to execute and use its output in the template
    is_command: Option<bool>,
}

#[cfg(test)]
mod testing {
    use crate::core::dates_names::{DailyName, MonthlyName, yearly_name::YearlyName};

    use super::*;
    #[test]
    fn should_resolve_placeholders_with_daily() {
        const TEST_INPUT: &str = r#"
[[placeholders]]
key = "hello"
value = "world"
[[placeholders]]
key = "no_command"
value = "pls_no_execute"
is_command=false 
[[placeholders]]
key = "command"
value = "echo hello"
is_command=true
[[placeholders]]
key = "day"
value = "{{DAY_JOURNAL}}"
[[placeholders]]
key = "month"
value = "{{MONTH_JOURNAL}}"
[[placeholders]]
key = "year"
value = "{{YEAR_JOURNAL}}"
"#;
        // Set up
        let config: AppConfig = toml::from_str(TEST_INPUT).expect("Invalid input from test input");

        let daily = DailyName::new(2000, 5, 15, crate::core::constants::MD_EXT).unwrap();
        let actual = act(&config, &daily);

        // Assert
        insta::assert_debug_snapshot!(actual);
    }

    #[test]
    fn should_resolve_placeholders_with_monthly() {
        const TEST_INPUT: &str = r#"
value = "{{DAY_JOURNAL}}"
[[placeholders]]
key = "month"
value = "{{MONTH_JOURNAL}}"
[[placeholders]]
key = "year"
value = "{{YEAR_JOURNAL}}"
"#;
        // Set up
        let config: AppConfig = toml::from_str(TEST_INPUT).expect("Invalid input from test input");

        let monthly = MonthlyName::from_ym(1980, 2, crate::core::constants::MD_EXT).unwrap();
        let actual = act(&config, &monthly);

        // Assert
        insta::assert_debug_snapshot!(actual);
    }
    #[test]
    fn should_resolve_placeholders_with_yearly() {
        const TEST_INPUT: &str = r#"
value = "{{DAY_JOURNAL}}"
[[placeholders]]
key = "month"
value = "{{MONTH_JOURNAL}}"
[[placeholders]]
key = "year"
value = "{{YEAR_JOURNAL}}"
"#;
        // Set up
        let config: AppConfig = toml::from_str(TEST_INPUT).expect("Invalid input from test input");

        let monthly = YearlyName::new(412.try_into().unwrap());
        let actual = act(&config, &monthly);

        // Assert
        insta::assert_debug_snapshot!(actual);
    }

    fn act<'a>(
        config: &'a AppConfig,
        journal: &'a impl ResolvePlaceholders,
    ) -> Vec<(&'a str, PlaceholderTemplate<'a, OsCommandProcossor>)> {
        let actual = config.create_template_placeholder_for(journal);

        // Prepare for assert
        let mut actual_as_vec: Vec<(&str, PlaceholderTemplate<'_, OsCommandProcossor>)> =
            actual.into_iter().collect();
        actual_as_vec.sort_by_key(|key_value| key_value.0);
        actual_as_vec
    }
}
