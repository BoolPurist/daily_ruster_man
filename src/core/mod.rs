mod daily_filtering;
mod daily_names;
use crate::AppResult;
pub mod data_models;
use data_models::*;
mod file_access;
mod process_handling;
use self::daily_names::DailyName;
pub fn open_today() -> AppResult {
    let today_name = DailyName::create_today_name();
    let to_open = file_access::create_new_path_for(today_name.name());

    process_handling::start_process_with(&to_open);
    Ok(())
}

pub fn fetch_all_daily_names() -> Vec<String> {
    let filter_by_valid_format = fetch_valid_daily_entries();
    filter_by_valid_format
        .into_iter()
        .map(|str| str.to_string())
        .collect()
}

fn fetch_valid_daily_entries() -> Vec<DailyName> {
    let daily_paths = file_access::get_all_daily_paths();

    let file_names = daily_filtering::strip_expect_file_name(&daily_paths);

    daily_filtering::filter_out_non_daily(file_names).collect()
}

pub fn open_by_future_past_range(range: &PastFuture) -> AppResult {
    let wanted_daily = DailyName::create_from_range(range);

    let to_open = file_access::create_new_path_for(wanted_daily.name());
    process_handling::start_process_with(&to_open);

    Ok(())
}
pub fn open_by_day_of_year(day_of_year: &DayOfYear) -> AppResult {
    let wanted_daily = DailyName::create_from_ordinal_day(day_of_year)?;

    let to_open = file_access::create_new_path_for(wanted_daily.name());
    process_handling::start_process_with(&to_open);

    Ok(())
}
pub fn open_by_year_month_day(ymd: &DayMonthYear) -> AppResult {
    let wanted_daily = DailyName::creaet_from_year_month_day(ymd)?;

    let to_open = file_access::create_new_path_for(wanted_daily.name());
    process_handling::start_process_with(&to_open);

    Ok(())
}
