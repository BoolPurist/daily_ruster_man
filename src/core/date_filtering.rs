use crate::core::{
    dates_names::{HasYear, DailyName, MonthlyName, HasMonth},
    date_models::find_by::{FindByMonthInYear, FindByYearMonthDay},
};
use std::path::Path;

#[cfg(test)]
#[path = "test_daily_filtering.rs"]
mod testing;

/// # Errors
/// - If path does not end with file name
/// - If file_name is not valid utf8
pub fn strip_expect_file_name<T>(paths: &[T]) -> impl Iterator<Item = &str>
where
    T: AsRef<Path>,
{
    let file_names_iter = paths.iter().filter_map(|full_path| {
        let full_path_ref = full_path.as_ref();
        let file_name = match full_path_ref.file_name() {
            Some(name) => name,
            None => {
                warn!(
                    "Could not get file name from path {full_path_ref:?} because it might end .. or is not file"
                );
                return None;
            }
        };

        match file_name.to_str() {
            Some(valid) => Some(valid),
            None => {
                warn!("Given file name {file_name:?} is not valid utf8");
                None
            }
        }
    });

    file_names_iter
}

pub fn filter_dailies_by_ymd(
    to_filter: Vec<DailyName>,
    ymd_listing: &FindByYearMonthDay,
) -> Vec<DailyName> {
    type FilterDate = (fn(&DailyName, u32) -> bool, u32);

    let mut filters: Vec<FilterDate> = Vec::with_capacity(3);

    if let Some(year) = ymd_listing.year() {
        filters.push((filter_by_year, year.into()));
    }
    if let Some(month) = ymd_listing.month() {
        filters.push((filter_by_month, month.into()));
    }
    if let Some(day) = ymd_listing.day() {
        filters.push((filter_by_day, day.into()));
    }

    filter_date_entries_by(to_filter, filters)
}

pub fn filter_monthly_by_ym(
    to_filter: Vec<MonthlyName>,
    ym_find_by: &FindByMonthInYear,
) -> Vec<MonthlyName> {
    type MonthlyFiltering = (fn(&MonthlyName, u32) -> bool, u32);
    let mut filters: Vec<MonthlyFiltering> = Vec::with_capacity(2);

    match ym_find_by {
        FindByMonthInYear::All => (),
        FindByMonthInYear::InCurrentYear(year) => {
            let year: u32 = (*year).into();
            filters.push((filter_by_year, year))
        }
        FindByMonthInYear::MonthYear { month, year } => {
            filters.push((filter_by_month, (*month).into()));
            filters.push((filter_by_year, (*year).into()));
        }
    }

    filter_date_entries_by(to_filter, filters)
}

fn filter_date_entries_by<T, FN>(to_filter: Vec<T>, filters: Vec<(FN, u32)>) -> Vec<T>
where
    FN: Fn(&T, u32) -> bool,
{
    to_filter
        .into_iter()
        .filter(|next| {
            filters.iter().all(|next_filter| {
                let (fnc, comp) = next_filter;
                fnc(next, *comp)
            })
        })
        .collect()
}

fn filter_by_year(to_check: &impl HasYear, year: u32) -> bool {
    to_check.is_in_year(year)
}
fn filter_by_month(to_check: &impl HasMonth, month: u32) -> bool {
    to_check.is_in_month(month)
}
fn filter_by_day(to_check: &DailyName, day: u32) -> bool {
    to_check.is_in_day(day)
}
