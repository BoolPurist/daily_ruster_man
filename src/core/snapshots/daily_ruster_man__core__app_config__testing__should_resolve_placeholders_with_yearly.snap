---
source: src/core/app_config.rs
expression: actual
---
[
    (
        "month",
        DirectValue(
            "{{MONTH_JOURNAL}}",
        ),
    ),
    (
        "year",
        DirectValue(
            "412",
        ),
    ),
]
